### PR TITLE
Added mechanism for generate demo data

### PR DESCRIPTION
**Pull request by ar4s**
I added new CLI command. Now, if you want to generate some example data use
``ralph make_demo_data``.
Available params for command:
 - --flush - flush the databases,
 - -d - execute fixture.

Example of use:
 - ``ralph make_demo_data --flush -d envs -d services`` - flush databases and execute defined fixtures
 - ``ralph make_demo_data --flush`` - interactive mode.

### CODE REVIEW

File: src/ralph/util/demo/__init__.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+from collections import namedtuple, defaultdict
+
+registry = {}
+demo_data = defaultdict(list)
+
+
+class DemoData(object):
+    def execute(self):
+        print('Create {}.'.format(self.name))
```
Comment (kula1922) [inline src/ralph/util/demo/__init__.py]: use logger

File: src/ralph/util/demo/devices.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+@register
+class DemoDeviceEnvironment(DemoData):
+    name = 'env'
+    title = 'Environments'
+
+    def generate_data(self, demo_data):
+        return {
+            'prod': DeviceEnvironmentFactory.create(
+                # type__id=11,
```
Comment (kula1922) [inline src/ralph/util/demo/devices.py]: uncomment ID

File: src/ralph/util/management/commands/make_demo_data.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            action='store_true',
+            dest='flush',
+            default=False,
+            help='Flush database before generate.'
+        ),
+    )
+
+    def handle(self, *args, **options):
+        flush = options.get('flush')
+        verbosity = options.get('verbosity')
```
Comment (kula1922) [inline src/ralph/util/management/commands/make_demo_data.py]: verbosity is not in option_list list

File: src/ralph/util/management/commands/make_demo_data.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        interactive = options.get('interactive')
+        if flush:
+            self.flush(verbosity)
+
+        if interactive:
+            selected = self.select_demos()
+            if not selected:
+                return None
+        else:
+            pass
```
Comment (kula1922) [inline src/ralph/util/management/commands/make_demo_data.py]: pass == return None. Just needless else statement

File: src/ralph/util/management/commands/make_demo_data.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        from ralph.util.demo import registry
+        for pos, demo in enumerate(registry):
+            self.stdout.write('{}) {}\n'.format(pos, demo.title))
+        choices = input('Please select demo data (eg.: 1, 4, 6): ')
+
+        if isinstance(choices, int):
+            registry[choices]().execute()
+        else:
+            for idx in choices:
+                registry[idx]().execute()
```
Comment (kula1922) [inline src/ralph/util/management/commands/make_demo_data.py]: for idx in raw_inout(Please select demo data (eg.: 1, 4, 6): ).split(',')
...

File: src/ralph/util/management/commands/make_demo_data.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        flush = options.get('flush')
+        verbosity = options.get('verbosity')
+        interactive = options.get('interactive')
+        if flush:
+            self.flush(verbosity)
+
+        if interactive:
+            selected = self.select_demos()
+            if not selected:
+                return None
```
Comment (kula1922) [inline src/ralph/util/management/commands/make_demo_data.py]: if not self.select_demos()

File: src/ralph/util/management/commands/make_demo_data.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+import textwrap
+import operator
+
+from django.core.management import call_command
+from django.core.management.base import BaseCommand
+from optparse import make_option
+
+
+class Command(BaseCommand):
+
```
Comment (kula1922) [inline src/ralph/util/management/commands/make_demo_data.py]: rm

File: src/ralph/util/demo/__init__.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        demo_data[self.name] = self.generate_data(demo_data)
+        print('Finish created {}.'.format(self.name))
+
+    def generate_data(self, demo_data):
+        raise NotImplementedError('Please override')
+
+
+def register(demo_klass, **kwargs):
+    # TODO: unikalna nazwa klucza w rejestrze
```
Comment (kula1922) [inline src/ralph/util/demo/__init__.py]: english :)

File: src/ralph/util/management/commands/make_demo_data.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    def handle(self, *args, **options):
+        flush = options.get('flush')
+        verbosity = options.get('verbosity')
+        interactive = options.get('interactive')
+        if flush:
+            self.flush(verbosity)
+        if interactive:
+            selected = self.select_demos()
+            if not selected:
+                return None
```
Comment (kula1922) [inline src/ralph/util/management/commands/make_demo_data.py]: if interactive:
    self.select_demos()
return None

File: src/ralph/util/demo/__init__.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    @property
+    def name():
+        raise NotImplementedError('Please specified name')
+
+    @property
+    def title():
+        raise NotImplementedError('Please specified title')
+
+    def execute(self):
+        print('Create {}.'.format(self.name))
```
Comment (kula1922) [inline src/ralph/util/demo/__init__.py]: use logger

File: src/ralph/util/demo/devices.py [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+@register
+class DemoDeviceEnvironment(DemoData):
+    name = 'envs'
+    title = 'Environments'
+
+    def generate_data(self, data):
+        # CITypes loaded from fixtures
+        ci_type = CIType.objects.get(id=11)
+        return {
+            'prod': DeviceEnvironmentFactory.create(
```
Comment (mkurek) [inline src/ralph/util/demo/devices.py]: wouldn't dev, prod and test be enough?

Comment (mkurek) [inline src/ralph/util/demo/devices.py]: btw what do you think about such a construction - wouldn't it be better?
```
envs = [('prod', 'Prod'), ('test', 'Test'), ('dev', 'Dev')]
return {slug: DeviceEnvironmentFactory(name=name, type=ci_type) for (slug, name) in envs.iteritems()}
```
eventually you could add some parser from name to slug (ex. `name.replace(' ', '_').lower()` in basic version) and just define list of names.

File: src/ralph/util/demo/devices.py [hunk 12]
```diff
@@ -0,0 +1,319 @@
...
+                type=ci_type,
```
Comment (mkurek) [inline src/ralph/util/demo/devices.py]: shouldn't type be defined in factory?

**Comment by vi4m** [comment]
Zajebioza(TM)

File: src/ralph/util/demo/core/cmdb.py [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+
+
+@register
+class DemoRelations(DemoData):
+    name = 'relations'
+    title = 'Relations'
+    required = ['services', 'envs']
+
+    def generate_data(self, data):
+        CIRelationFactory.create(
```
Comment (vi4m) [inline src/ralph/util/demo/core/cmdb.py]: why not for loop?

File: src/ralph/util/management/commands/make_demo_data.py [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+            '--flush',
+            action='store_true',
+            dest='flush',
+            default=False,
+            help='Flush database before generate.'
+        ),
+        make_option(
+            '--fixtures',
+            type='string',
+            action='callback',
```
Comment (mkurek) [inline src/ralph/util/management/commands/make_demo_data.py]: there is no need to parse this list by hand (and use setattr :hear_no_evil:). Use sth like this:
```
make_option(
    '--fixtures',
    dest='fixtures',
    action='append',
    type='str',
    default=[],
    help=_('Fixtures to apply'),
)
```
Then you can call `ralph make_demo_data --fixtures fix1 fix2 fix3`

Comment (ar4s) [inline src/ralph/util/management/commands/make_demo_data.py]: Thanks! I'm noob in ``optparse`` :crying_cat_face:

File: src/ralph/util/demo/__init__.py [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+registry = {}
+demo_data = defaultdict(dict)
+
+
+class DemoData(object):
+    required = None
+
+    @property
+    def name():
+        raise NotImplementedError('Please specified name')
```
Comment (vi4m) [inline src/ralph/util/demo/__init__.py]: s/specified/specify

File: src/ralph/util/demo/__init__.py [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+        super(DemoRunner, self).__init__(*args, **kwargs)
+
+    def run(self):
+        """This method find reqiurements of demo data and call execute method
+        for each demo in demo_keys."""
+        demos = Counter()
+        max_depth = 10
+        default_weight = 10
+
+        def dig_requirements(item, depth=1):
```
Comment (vi4m) [inline src/ralph/util/demo/__init__.py]: should i *really* read the function body just to know what is this doing?

**Comment by vi4m** [comment]
One line documentation in the Installation instruction.

File: src/ralph/util/demo/__init__.py [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+
+
+class DemoRunner(object):
+    """Simple helper class for collection of demo data."""
+    def __init__(self, demo_keys, *args, **kwargs):
+        self.demo_keys = demo_keys
+        super(DemoRunner, self).__init__(*args, **kwargs)
+
+    def run(self):
+        """This method find reqiurements of demo data and call execute method
```
Comment (vi4m) [inline src/ralph/util/demo/__init__.py]: requirement

File: src/ralph/util/management/commands/make_demo_data.py [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+            selected = self.select_demos()
+            if not selected:
+                return
+            demo = DemoRunner(selected)
+        else:
+            demo = DemoRunner(fixtures)
+        if demo:
+            demo.run()
+
+    def select_demos(self):
```
Comment (vi4m) [inline src/ralph/util/management/commands/make_demo_data.py]: docstring missing. What is select demos?

Comment (vi4m) [inline src/ralph/util/management/commands/make_demo_data.py]: Please kindly provide documentation for idea behind the source code.

File: src/ralph/util/demo/core/discovery.py [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+            'b': DeprecatedDataCenterFactory(
+                name=data['discovery_queue']['b'].name,
+                service=data['services']['infrastructure'],
+                device_environment=data['envs']['prod'],
+            ),
+        }
+
+
+@register
+class DemoDeprecatedRack(DemoData):
```
Comment (vi4m) [inline src/ralph/util/demo/core/discovery.py]: why data center rack doesn't add missing deprecated rack into the ralph?

**Comment by vi4m** [comment]
1. Why we need --flush? 2. How to add all demodata easily?

**Comment by vi4m** [comment]
Change Flush -> purge.
Default behavior is to make all demo data.
make -i explicit.

**Comment by vi4m** [comment]
Please make another pull request with the fixes - i will merge this for now.

**Pull request merged**